### ISSUE TITLE
auth: record types besides CNAME revealed in NSEC3 record

### ISSUE THREAD

**Issue by peterthomassen** [PowerDNS/pdns#5400]
- Program: Authoritative
 - Issue type: Bug report

### Short description
According to RFC 1912, "A CNAME record is not allowed to coexist with any other data". According to my tests, pdns seems to handle this correctly by hiding other records if a CNAME is present, only responding with the CNAME.

Now, the NSEC3 record of the "previous" (in `ordername` sense) name is supposed to list the record types that are available. This record does list other record types present in the database.

As far as I understand, the correct behavior would be to list only CNAME and RRSIG. (I suppose that NSEC records are also affected, but did not test.)

### Environment

 - Operating system: Ubuntu
 - Software version: 4.0.3-1pdns.jessie
 - Software source: your repository

### Steps to reproduce

1. Create a zone with DNSSEC and NSEC3 turned on
2. Create a CNAME record for some name of said zone, along with other records for the same name
3. Inspect the NSEC3 record of the "previous" (ordername-sense) name

### Expected behaviour
When creating `cname.example2.dedyn.io` with CNAME, TXT, and A records, `cnam.example2.dedyn.io`'s NSEC3 should only mention the CNAME and RRSIG record types.

### Actual behaviour
When creating `cname.example2.dedyn.io` with CNAME, TXT, and A records, `cnam.example2.dedyn.io`'s NSEC3 record turns out to be:

    083nl339nqu4vm32trdieakq0the5g53.example2.dedyn.io. 60 IN NSEC3	1 0 300 A034DC3BEBA750B1 ETKJ0J7PHTBD7OLUIGSQKRNJRKA3GPC1 A CNAME TXT RRSIG

... with A and TXT being listed along with CNAME, RRSIG.

**Comment by mind04** [comment]
Powerdns is not hiding other record types. Behavior depend on the record order in the DB.
It is incredibly expensive to check for cname vs record type for every query. It will more or less double the db query load. And axpex (domain name without additional labels) will always be a problem.

Clean your data in the db to avoid this and validate input in you frontend. 'pdndutil check-all-zones' might be of help to locate current problem zones.

Please don't fix crap in the db in powerdns, since this will only lead to more/better crap. This is a battle you cannot win.

**Comment by Habbie** [comment]
I'm afraid I will have to agree with @mind04 on this. Closing ticket, sorry.

**Comment by peterthomassen** [comment]
I generally agree that it is the best approach to keep the database consistent, and I don't think the issue I described necessarily needs to be fixed.

However, I would like to point out in response to @mind04's comment that what's responsible for finding e.g. an `A` record value is the [`basic-query` setting](https://github.com/PowerDNS/pdns/blob/master/modules/gmysqlbackend/gmysqlbackend.cc#L81), which explicitly selects from the `records` table with a specific `name` and `type` requirement. As a consequence, the relative order of `CNAME` and `A` records in the database does not matter. Furthermore, pdns does check for the presence of a `CNAME` record, as described [in the blog](https://blog.powerdns.com/2015/06/23/what-is-a-powerdns-backend-and-how-do-i-make-it-send-an-nxdomain/).

Now, the result of this is that if DNSSEC is turned off, pdns serves compliant responses, even if the database has other stuff for the same name, which is then ignored. Once you turn on DNSSEC, you end up with inconsistent `NSEC3` records. Because in the non-DNSSEC case, things work fine, this may not be the expected behavior. (It's not a complaint, just stating more clearly why I posted this issue.)
